service/biz/enrollz_biz.go

```diff
@@ -17,27 +17,123 @@ package biz
+	RotateOIakCert(req *epb.RotateOIakCertRequest) (*epb.RotateOIakCertResponse, error)
+}
+
+// Infra-specific dependencies of this enrollz business logic lib. A service can create all these dependencies
+// and wire them to the library on server start-up.
+type EnrollzInfraDeps interface {
+	// Client to communicate with Switch Owner CA to issue oIAK and oIDevID certs.
+	SwitchOwnerCaClient
+
+	// Client to communicate with the switch's enrollz endpoints.
+	EnrollzDeviceClient
+}
+
+// This is a "client"/switch-owner side implementation of Enrollz service. This client-side logic
+// is part of the switch owner infra/service and is expected to communicate with a device/switch (hosting
+// enrollz gRPC endpoints) to verify its TPM-based identity and provision it with the switch-owner-issued,
+// TPM-based attestation and TLS certs: Owner IAK and Owner IDevID certs respectively. Switch owner is
+// expected to TPM-enroll one switch control card at a time, starting with an active card.
+func EnrollControlCard(controlCardSelection *cpb.ControlCardSelection, deps EnrollzInfraDeps) error {
 	// 1. Fetch Switch Vendor CA trust bundle.
 
-	// 2. Call device's GetIakCert API for Active control card
+	// 2. Call device's GetIakCert API for the specified control card.
+	getIakCertReq := &epb.GetIakCertRequest{ControlCardSelection: controlCardSelection}
+	getIakCertResp, err := deps.GetIakCert(getIakCertReq)
+	if err != nil {
+		log.Errorf("Error retrieving IAK cert from the device with req=%s: %v",
+			prototext.Format(getIakCertReq), err)
+		return err
+	}
+	log.Infof("Successfully received from device GetIakCert() resp=%s for req=%s",
+		prototext.Format(getIakCertResp), prototext.Format(getIakCertReq))
+
+	// 3. Validate IDevID TLS cert if control card is standby (for active card IDevID would
+	// be validated during the TLS handshake).
+
+	// 4. Validate IAK cert.
 
-	// 3. Validate IDevID TLS cert
+	// 5. Make sure IAK and IDevID serials match.
 
-	// 4. Validate IAK cert
+	// 6. Extract IAK pub from IAK cert and validate it (accepted crypto algo and key length).
 
-	// 5. Make sure IAK and IDevID serials match
+	// 7. Extract IDevID pub from IDevID cert and validate it (accepted crypto algo and key length).
 
-	// 6. Extract IAK pub from IAK cert and validate it (accepted crypto algo and key length)
+	// 8. Call Switch Owner CA to issue oIAK and oIDevID certs.
 
-	// 7. Extract IDevID pub from IDevID cert and validate it (accepted crypto algo and key length)
+	// Pass IAK cert PEM as is for now, but use IAK pub key PEM instead once x509 cert parser
+	// logic is implemented.
+	oIakCertPem, err := deps.IssueOwnerIakCert(getIakCertResp.ControlCardId, getIakCertResp.IakCert)
+	if err != nil {
+		log.Errorf("Error executing Switch Owner CA IssueOwnerIakCert() with control_card_id=%s IAK_pub_pem=%s: %v",
+			prototext.Format(getIakCertResp.ControlCardId), getIakCertResp.IakCert, err)
+		return err
```

Comment:
Nit: In the code I've seen, usually errors are returned wrapped in an error message:

return fmt.Errorf("Error executing Switch Owner CA IssueOwnerIakCert() with control_card_id=%s IAK_pub_pem=%s: %v", prototext.Format(getIakCertResp.ControlCardId), getIakCertResp.IakCert, err)

That way instead of log.Errorf many times in each function, you can just log any errors from the parent function. It's arguable whether doing it this way is actually better but I'm also more familiar with C++ so idk

go/go-style/best-practices#error-structure

Reply:
Gotcha and thanks for the pointer. I think it makes sense to return an error wrapper with more details back to callers and let them handle it.